File: frontend/src/lexer/token.rs
use std::fmt::Display;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Token {
    EOF,
    Id(String),
    Num(u64),
    Assign,    // `=`
    Define,    // `:=`
    Eq,        // `==`
    NotEq,     // `!=`
    And,       // `&&`
    Or,        // `||`
    Less,      // `<`
    LessEq,    // `<=`
    Greater,   // `>`
    GreaterEq, // `>=`
    Arrow,     // `->`
    Fn,        // `fn`
    Type,      // `type`
    Mut,       // `mut`
    Ret,       // `ret`
    True,      // `true`
    False,     // `false`
    For,       // `for`
    Do,        // `do`
    If,        // `if`
    Then,      // `then`
    Else,      // `else`
    Struct,    // `struct`
    EOL,       //
    LParen,    // `(`
    RParen,    // `)`
    LBrace,    // `{`
    RBrace,    // `}`
    Quote,     // `"`
    Dot,       // `.`
    Comma,     // `,`
    Colon,     // `:`
    Plus,      // `+`
    Minus,     // `-`
    Mul,       // `*`
    Exclam,    // `!`
    Amper,     // `&`
    Vbar,      // `|`
}

impl Token {
    pub fn to_str(&self) -> &'static str {
        match self {
            Token::EOF => "end of input",
            Token::Id(_) => "identifier",
            Token::Num(_) => "number",
            Token::Assign => "`=`",
            Token::Define => "`:=`",
            Token::Arrow => "`->`",
            Token::Eq => "`==`",
            Token::NotEq => "`!=`",
            Token::And => "`&&`",
            Token::Or => "`||`",
            Token::Less => "`<`",
            Token::LessEq => "`<=`",
            Token::Greater => "`>`",
            Token::GreaterEq => "`>=`",
            Token::Fn => "`fn`",
            Token::Type => "`type`",
            Token::Mut => "`mut`",
            Token::Ret => "`ret`",
            Token::True => "`true`",
            Token::False => "`false`",
            Token::For => "`for`",
            Token::Do => "`do`",
            Token::If => "`if`",
            Token::Then => "`then`",
            Token::Else => "`else`",
            Token::Struct => "`struct`",
            Token::EOL => "end of line",
            Token::LParen => "`(`",
            Token::RParen => "`)`",
            Token::LBrace => "`{`",
            Token::RBrace => "`}`",
            Token::Quote => "`\"`",
            Token::Dot => "`.`",
            Token::Comma => "`,`",
            Token::Colon => "`:`",
            Token::Plus => "`+`",
            Token::Minus => "`-`",
            Token::Mul => "`*`",
            Token::Exclam => "`!`",
            Token::Amper => "`&`",
            Token::Vbar => "`|`",
        }
    }
}

impl Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.to_str())
    }
}

File: src/main/scala/frontend/lexer/Token.scala
package nop.frontend.lexer;

enum Token:
  case EOF
  case Id(val s: String)
  case Num(val i: Int)
  case Assign    // `=`
  case Define    // `:=`
  case Eq        // `==`
  case NotEq     // `!=`
  case And       // `&&`
  case Or        // `||`
  case Less      // `<`
  case LessEq    // `<=`
  case Greater   // `>`
  case GreaterEq // `>=`
  case Arrow     // `->`
  case Fn        // `fn`
  case Type      // `type`
  case Mut       // `mut`
  case Ret       // `ret`
  case True      // `true`
  case False     // `false`
  case For       // `for`
  case Do        // `do`
  case If        // `if`
  case Then      // `then`
  case Else      // `else`
  case Struct    // `struct`
  case EOL       //
  case LParen    // `(`
  case RParen    // `)`
  case LBrace    // `{`
  case RBrace    // `}`
  case Dot       // `.`
  case Comma     // `,`
  case Colon     // `:`
  case Plus      // `+`
  case Minus     // `-`
  case Mul       // `*`
  case Exclam    // `!`
  case Amper     // `&`
  case Vbar      // `|`
